# Module 3 Lab 2
Screen a real crime dataset for values that cannot be true, and practice the analyst’s distinction between impossible and merely extreme.

Tasks:
    Import data
      Address missing values
      Look for outliers and smooth noisy data
      Prepare the dataset to establish a relation between an urban population category and crime type.
         (see textbook to define discreet categories)


In [1]:
import pandas as pd
originaldf = pd.read_csv('usarrests.csv')
df = originaldf

In [2]:
df.shape

(50, 4)

In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  50 non-null     str    
 1   Murder      50 non-null     float64
 2   Assault     49 non-null     float64
 3   UrbanPop    50 non-null     int64  
dtypes: float64(2), int64(1), str(1)
memory usage: 2.1 KB


In [4]:
df.head()

,Unnamed: 0,Murder,Assault,UrbanPop
0,Alabama,13.2,236.0,58
1,Alaska,10.0,263.0,48
2,Arizona,8.1,294.0,80
3,Arkansas,8.8,190.0,50
4,California,9.0,276.0,91


In [5]:
df.describe()

,Murder,Assault,UrbanPop
count,50.00000,49.000000,50.00000
mean,7.78800,182.183673,74.20000
std,4.35551,130.877435,73.40828
min,0.80000,45.000000,6.00000
25%,4.07500,109.000000,53.25000
50%,7.25000,159.000000,66.00000
75%,11.25000,249.000000,77.75000
max,17.40000,879.000000,570.00000


In [6]:
df.isnull().sum()

Unnamed: 0    0
Murder        0
Assault       1
UrbanPop      0
dtype: int64

In [7]:
print(df.duplicated().sum())

0


# Initial Data Summary

The data set is (50,4)
Includes columns: Unnamed 0, Murder, Assault, UrbanPop for all 50 states (Unnamed is States column)

These columns correspond to arrests per 100,000 residents for assault and murder and the percentage of the state's population living in urban areas in 1973. 

Problems to address:
+ 1 assault value is missing, Georgia
+ No duplicates
+ No negative values (as seen in min)
+ Dtypes look good
+ Apparent outlier in UrbanPop, Iowa (570 is max value in data set, not possible, 100 is max allowed for UrbanPop %) May be a decimal error.
+ UrbanPop New York is 6. Very low. Others between 32 and 91. This is extreme, may be a decimal error.
+ South Carolina Assault rate is very high, 879 others between 45 and 337. This is extreme, may be a decimal error.
+ Apparent Outlier in Murder, min is 0.8 (North Dakota), all other values between 2 and 17.4. This is extreme, may be a decimal error.
+ Unnamed column could be named "States"

# Impossible Value: UrbanPop, Iowa 570 to 57.0. 
Value is impossible, 100 is max allowed for UrbanPop %. Assume decimal error. Other values between 32 and 91. Change to 57.0.

In [8]:
df.sort_values(["UrbanPop"])

,Unnamed: 0,Murder,Assault,UrbanPop
31,New York,11.1,254.0,6
44,Vermont,2.2,48.0,32
47,West Virginia,5.7,81.0,39
33,North Dakota,0.8,45.0,44
23,Mississippi,16.1,259.0,44
40,South Dakota,3.8,86.0,45
32,North Carolina,13.0,337.0,45
1,Alaska,10.0,263.0,48
39,South Carolina,14.4,879.0,48
3,Arkansas,8.8,190.0,50


Replace row 14 UrbanPop with 57

In [9]:
df.loc[14, "UrbanPop"] = 57

In [10]:
df.head(15)

,Unnamed: 0,Murder,Assault,UrbanPop
0,Alabama,13.2,236.0,58
1,Alaska,10.0,263.0,48
2,Arizona,8.1,294.0,80
3,Arkansas,8.8,190.0,50
4,California,9.0,276.0,91
5,Colorado,7.9,204.0,78
6,Connecticut,3.3,110.0,77
7,Delaware,5.9,238.0,72
8,Florida,15.4,335.0,80
9,Georgia,17.4,NaN,60


# Outlier changes and flags
+ UrbanPop New York is 6. Very low. Others between 32 and 91. Assume decimal error and look up statistics on internet. Changed New York UrbanPop from 6 to 60
+ South Carolina Assault rate is very high, 879 others between 45 and 337. This is extreme. However Google search indicates high rates of assault in SC, so data is flagged, but not changed.
+ Apparent Outlier in Murder minimum (0.8  North Dakota), all other values between 2 and 17.4. Google search indicates very low rates for ND, so data is flagged, but not changed.


In [11]:
df.loc[31, "UrbanPop"] = 60

In [12]:
print(df)

        Unnamed: 0  Murder  Assault  UrbanPop
0          Alabama    13.2    236.0        58
1           Alaska    10.0    263.0        48
2          Arizona     8.1    294.0        80
3         Arkansas     8.8    190.0        50
4       California     9.0    276.0        91
5         Colorado     7.9    204.0        78
6      Connecticut     3.3    110.0        77
7         Delaware     5.9    238.0        72
8          Florida    15.4    335.0        80
9          Georgia    17.4      NaN        60
10          Hawaii     5.3     46.0        83
11           Idaho     2.6    120.0        54
12        Illinois    10.4    249.0        83
13         Indiana     7.2    113.0        65
14            Iowa     2.2     56.0        57
15          Kansas     6.0    115.0        66
16        Kentucky     9.7    109.0        52
17       Louisiana    15.4    249.0        66
18           Maine     2.1     83.0        51
19        Maryland    11.3    300.0        67
20   Massachusetts     4.4    149.

# Address missing value

In [13]:
rows_with_na = df[df.isna().any(axis=1)]
print(rows_with_na)

  Unnamed: 0  Murder  Assault  UrbanPop
9    Georgia    17.4      NaN        60


Let's look at mean Murder and Mean UrbanPop to see where Georgia falls. 


Otherwise, we can average the values from States with similar statistics as an option.

In [14]:
print(df["Murder"].mean())

7.788


In [15]:
print(df["UrbanPop"].mean())

65.02


Georgia has the highest murder rate, so not near the murder mean.
The urban population is less than the mean so these two values do not agree with repsect to guiding the decision.

In [16]:
df[(df["Murder"] > 15)]

,Unnamed: 0,Murder,Assault,UrbanPop
8,Florida,15.4,335.0,80
9,Georgia,17.4,NaN,60
17,Louisiana,15.4,249.0,66
23,Mississippi,16.1,259.0,44


There are only three other states with murder rates above 15. This is too small of a data set to average.

In [17]:
df.sort_values(["Murder", "Assault"])

,Unnamed: 0,Murder,Assault,UrbanPop
33,North Dakota,0.8,45.0,44
28,New Hampshire,2.1,57.0,56
18,Maine,2.1,83.0,51
44,Vermont,2.2,48.0,32
14,Iowa,2.2,56.0,57
48,Wisconsin,2.6,53.0,66
11,Idaho,2.6,120.0,54
22,Minnesota,2.7,72.0,66
43,Utah,3.2,120.0,80
6,Connecticut,3.3,110.0,77


In [18]:
df.sort_values(["UrbanPop", "Assault"])

,Unnamed: 0,Murder,Assault,UrbanPop
44,Vermont,2.2,48.0,32
47,West Virginia,5.7,81.0,39
33,North Dakota,0.8,45.0,44
23,Mississippi,16.1,259.0,44
40,South Dakota,3.8,86.0,45
32,North Carolina,13.0,337.0,45
1,Alaska,10.0,263.0,48
39,South Carolina,14.4,879.0,48
3,Arkansas,8.8,190.0,50
18,Maine,2.1,83.0,51


In [19]:
df[(df["Murder"] > 13)]

,Unnamed: 0,Murder,Assault,UrbanPop
0,Alabama,13.2,236.0,58
8,Florida,15.4,335.0,80
9,Georgia,17.4,NaN,60
17,Louisiana,15.4,249.0,66
23,Mississippi,16.1,259.0,44
39,South Carolina,14.4,879.0,48
41,Tennessee,13.2,188.0,59


There seems to be a better correlation between assault rates and murder rates rather than Urban pop and assautl. Therefore, I will average the assault rate for all Murder rates above 13.2 (6 values) and replace Georgia's missing murder value with that avaerage. 

Top 7 murder rate states data:
+ Murder rates: 13.2 to 17.4 (Georgia 17.4)
+ Urban Pop: 44 to 80 (Georgia 60)

Google prompt: how do I average the 7 highest values in a pandas dataframe?
Google answer: avg_highest = df['column_name'].nlargest(7).mean()
print(avg_highest)

In [20]:
avg_highest = df['Assault'].nlargest(7).mean()
print(avg_highest)

386.57142857142856


This mean value does include the anomoulously high SC rate, so may be high.

In [21]:
df.fillna(386.6)

,Unnamed: 0,Murder,Assault,UrbanPop
0,Alabama,13.2,236.0,58
1,Alaska,10.0,263.0,48
2,Arizona,8.1,294.0,80
3,Arkansas,8.8,190.0,50
4,California,9.0,276.0,91
5,Colorado,7.9,204.0,78
6,Connecticut,3.3,110.0,77
7,Delaware,5.9,238.0,72
8,Florida,15.4,335.0,80
9,Georgia,17.4,386.6,60


###### UrbanPop Categories

I will add a column that lists urban population categories.
+ small <50
+ medium 50-69
+ large >70

I do not know how to add a column.

Google prompt: How do I ass a string column in pandas?
Google answer: # Add a string column with a default value
df['Status'] = 'Pending'

If (UrbanPop)<50
    ["UrbanPop_size"] == "small"
    

In [22]:
df['UrbanPopSize'] = 'Pending'

In [23]:
df.head()

,Unnamed: 0,Murder,Assault,UrbanPop,UrbanPopSize
0,Alabama,13.2,236.0,58,Pending
1,Alaska,10.0,263.0,48,Pending
2,Arizona,8.1,294.0,80,Pending
3,Arkansas,8.8,190.0,50,Pending
4,California,9.0,276.0,91,Pending


# My Failed Code
for x in df["UrbanPop"]:
    if x < 50:
        df["UrbanPopSize"] == "small"
    elif x < 70:
        df["UrbanPopSize"] == "medium"
    else:
        df["UrbanPopSize"] == "large"


# Asked Gemini for help

Gemini answer:

def categorize_pop(val):
    if val < 50:
        return 'small'
    elif val < 70:
        return 'medium'
    else:
        return 'large'

df['UrbanPopSize'] = df['UrbanPop'].apply(categorize_pop)


In [24]:
def categorize_pop(val):
    if val < 50:
        return 'small'
    elif val < 60:
        return 'medium'
    elif val < 70:
        return 'large'
    else :
        return 'extra large'
    

df['UrbanPopSize'] = df['UrbanPop'].apply(categorize_pop)


In [25]:
df.head()

,Unnamed: 0,Murder,Assault,UrbanPop,UrbanPopSize
0,Alabama,13.2,236.0,58,medium
1,Alaska,10.0,263.0,48,small
2,Arizona,8.1,294.0,80,extra large
3,Arkansas,8.8,190.0,50,medium
4,California,9.0,276.0,91,extra large


In [26]:
df.sort_values(["UrbanPop"])

,Unnamed: 0,Murder,Assault,UrbanPop,UrbanPopSize
44,Vermont,2.2,48.0,32,small
47,West Virginia,5.7,81.0,39,small
33,North Dakota,0.8,45.0,44,small
23,Mississippi,16.1,259.0,44,small
40,South Dakota,3.8,86.0,45,small
32,North Carolina,13.0,337.0,45,small
1,Alaska,10.0,263.0,48,small
39,South Carolina,14.4,879.0,48,small
3,Arkansas,8.8,190.0,50,medium
18,Maine,2.1,83.0,51,medium


# Dataframe Cleaning Notes

A (50,4) dataframe was downloaded.

+ No duplicates
+ No negative values (as seen in min)
+ Dtypes look good

Flagged and addressed issues:
+ Impossible outlier in UrbanPop, Iowa 570 (%). 570 is not a possible percentage, 100 is max allowed for UrbanPop %. Assumed decimal error and changed value from 570 to 57.0.
+ UrbanPop New York is 6 (%). Very low. Others between 32 and 91. This is extreme. Checked google and assumed decimal error. Changed value from 6 to 60.
+ Replaced missing value for Georgia assault rate. Value was missing. Georgia had the highest murders and murders seeemed to correlate somewhat with assaults. I sorted out the 7 States with the highest murder rates. I averaged their assaults and used that average as the Georgia assaulty rate. NOTE: the assault rate for South Carolina was anamalously high and was used in this calculation. Thus the imputed value for Georgia's assault rate could be elevated.
  
+ Possible outlier - South Carolina Assault rate is very high, 879. All other values were between 45 and 337. This is extreme, may be a decimal error but could not confirm. Left as is with a flag that this may be high value.
+ Possible outlier -  North Dakota's Murder rate is very low at 0.8. All other values were between 2 and 17.4. This value is extreme, may be a decimal error but could not confirm. Left as is with a flag that this may be low value. 

In [27]:
df.to_csv("usarrests_clean.csv", index=False)

https://github.com/acamend/cs82a-portfolio